sw/cheri/common/platform-pinmux.hh.tpl

```diff
@@ -1,195 +1,201 @@
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
-   */
-  bool output_pin_select(OutputPin output_pin, uint8_t option) {
-    if (option >= output_pin_options(output_pin)) {
-      Debug::log("Selected option is not valid for this pin.");
-      return false;
-    }
-    uint16_t registerOffset   = static_cast<uint16_t>(output_pin);
-    registers[registerOffset] = (1 << option);
-    return true;
-  }
+/**
+ * A handle to a sink configuration register. This can be used to select
+ * the source of the handle's associated sink.
+ */
+template <typename SinkEnum>
+struct Sink {
+  CHERI::Capability<volatile uint8_t> reg;
+  const SinkEnum sink;
 
   /**
-   * For a given block input, selects a pin to use for that input via the pin
-   * multiplexer.
+   * Select a source to connect to the sink.
    *
-   * @param block_input The block input to pinmux.
-   * @param option The option to select for that block input. This value
-   * should be less than the value returned by `block_input_options` for the
-   * given block input.
+   * To see the sources available for a given sink see the Sonata system
+   * documentation:
+   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux/
    *
-   * The meanings of these selections can be found in the documentation:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
+   * Note, source 0 disconnects the sink from any source disabling it,
+   * and source 1 is the default source for any given sink.
    */
-  bool block_input_select(BlockInput block_input, uint8_t option) {
-    if (option >= block_input_options(block_input)) {
-      Debug::log("Selected option is not valid for this block.");
+  bool select(uint8_t source) {
+    if (source >= sources_number(sink)) {
+      Debug::log("Selected source not within the range of valid sources.");
       return false;
     }
-    uint16_t registerOffset   = static_cast<uint16_t>(block_input);
-    registers[registerOffset] = (1 << option);
+    *reg = 1 << source;
     return true;
   }
 
-  /**
-   * A constructor for the SonataPinmux driver, which takes a bounded
-   * capability to the pinmux registers. This should be replaced with
-   * an appropriate `MMIO_CAPABILITY` call in the version of the driver
-   * that runs in CHERIoT RTOS, rather than baremetal.
-   */
-  SonataPinmux(volatile uint8_t *registers) : registers(registers) {}
+  /// Disconnect the sink from all available sources.
+  void disable() { *reg = 0b01; }
+
+  /// Reset the sink to it's default source.
+  void default_selection() { *reg = 0b10; }
+};
+
+namespace {
+template <typename SinkEnum>
+// This is used by `BlockSinks` and `PinSinks`
+// to return a capability to a single sink's configuration register.
+inline Sink<SinkEnum> _get_sink(volatile uint8_t *base_register, const SinkEnum sink) {
+  CHERI::Capability reg = {base_register + static_cast<ptrdiff_t>(sink)};
+  reg.bounds()          = sizeof(uint8_t);
+  return Sink<SinkEnum>{reg, sink};
```

Comment:
There will be two contiguous MMIO regions described in the board file: block sinks and pin sinks. Any compartment could request access to either or both. This can be audited normally.

Yes, the idea of having `{Pin,Block}Sinks::get()` return a capability with a bound only covering the particular sink is that this allows a compartment to give another compartment permission to configure the particular sink, without giving access to the whole region.

Reply:
I see, that makes sense. I think it would be nicer if for auditing purposes no compartments were ever made to get access to the full MMIO regions, and instead only had to get access to the pins required, but I suppose that would either require defining 155 sink devices in the board description file, or (preferably) changes in upstream `cheriot-rtos` to support this kind of granular capability loading functionality (e.g. `MMIO_CAPABILITY(uint8_t, pinmux, offset=0x012, bounds=0x8)` would be nice for example).